subtests/docker_cli/systemd/systemd.py

```diff
@@ -0,0 +1,81 @@
+"""
+Summary
+-------
+Testing basic container as systemd service
+
+Operational Summary
+---------------------
+This is to test container as systemd service using a predefined unit file
+(p4321.service) and a script p4321-server.py. An image is built with this
+script added. Systemd starts a container using this image & the container
+writes current time to port 4321. From host, the socket is read and the value
+is checked to see if it is a digit. Finally, container is stopped through
+systemd and system is cleaned up.
+
+"""
+
+import shutil
+import time
+import socket
+from os.path import join
+from os.path import exists
+from os import unlink
+from autotest.client import utils
+from autotest.client.shared.utils import is_port_free
+from dockertest.subtest import Subtest
+from dockertest.dockercmd import DockerCmd
+from dockertest.output import mustpass
+
+
+class systemd(Subtest):
+
+    def initialize(self):
+        super(systemd, self).initialize()
+        self.stuff['systemd_touchstone'] = False
+        sysd_unitf = join(self.bindir, 'p4321.service')
+        shutil.copyfile(sysd_unitf, self.config['sysd_unitf_dest'])
+        dkrcmd = DockerCmd(self, 'build', ['--force-rm -t',
+                                           self.config['name'], self.bindir])
+        mustpass(dkrcmd.execute())
+
+    def run_once(self):
+        host = 'localhost'
+        port = 4321
+        self.sysd_action('daemon-reload')
+        self.sysd_action('start', self.config['sysd_unit_file'])
+        while is_port_free(port, host):
+            time.sleep(0.1)
+        time_from_socket = self.read_socket(host, port)
+        self.failif(not time_from_socket.isdigit(),
+                    "Data received from container is non-numeric: '%s'"
+                    % time_from_socket)
+        self.sysd_action('stop', self.config['sysd_unit_file'])
+
+    def sysd_action(self, action=None, sysd_unit=None):
+        self.stuff['systemd_touchstone'] = True
+        command = "systemctl " + action
+        if sysd_unit:
+            command += " " + sysd_unit
+        utils.run(command, ignore_status=False)
+
+    def cleanup(self):
+        super(systemd, self).cleanup()
+        filepath = self.config['sysd_unitf_dest']
+        if exists(filepath):
+            unlink(filepath)
+        if self.stuff['systemd_touchstone']:
+            self.sysd_action('daemon-reload')
+
+    @staticmethod
+    def read_socket(host, port):
+        time_from_socket = ''
+        num_read_try = 10
+        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
```

Comment:
So, couple of things. First, as you suggested, a `sleep` is necessary between retries. `sleep(0.5)` is IMHO a reasonable value.

Second, I think you have to have `socket.socket` inside the while loop, not outside. Retrying `recv()` after EOF doesn't seem to work; yesterday I saw cases where just retrying `recv` just kept getting empty string. I haven't reproduced it today, though.

This is what I have working on my end; but note that it doesn't incorporate @cevich's changes because I haven't taken time to look at `create_connection` or `settimeout`.

def read_socket(host, port):
        num_read_try = 10
        while num_read_try:
            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            s.connect((host, port))
            time_from_socket = s.recv(16).strip()
            if time_from_socket:
                s.close()
                return time_from_socket
            self.logwarning("empty read from socket")   # just for my debugging
            num_read_try -= 1
            time.sleep(0.5)
        raise SomeSortOfExceptionIndicatingThatWeGotNoDataFromSocket